### PR TITLE
[tests] Add GRU quantization test

### PR DESCRIPTION
**Pull request by AnthonyOfSeattle**
Builds an explicitly initialized 2 GRU graph and tests quantization identically to end2end test. Adds a single TopKNode to make sure some quantization test coverage is available for that node type.

Addresses #985 by increasing test coverage in glow/lib/Quantization.cpp

Possible issues:
Runs with following warning on CPU backend
```
remark: /home/anthony/glow/lib/Backends/CPU/libjit/libjit.cpp:677:13: loop not vectorized: call instruction cannot be vectorized
warning: <unknown>:0:0: loop not vectorized: failed explicitly specified loop vectorization
```

### CODE REVIEW

**Comment by rdzhabarov** [comment]
Thanks, @AnthonyOfSeattle for contributing. Can you please sign the CLA above, that's required in order to merge your changes. Once that's done, we'll review the PR.

File: tests/unittests/quantizationTest.cpp [hunk 1]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+    float mx = result2Handle.raw(result2Handle.minMaxArg().second);
+    double diff = std::fabs(result2Handle.raw(i) - result1Handle.raw(i)) / mx;
 
+    // Allow 3% difference.
+    EXPECT_NEAR(diff, 0, 0.03);
+  }
+}
+
+/// Fills the tensor \p H with some stable random integers with the seed \p seed
+/// and the range [0 .. scale].
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: [0, scale)

File: tests/unittests/quantizationTest.cpp [hunk 2]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+static void fillStableRandomIndex(Handle<size_t> H, size_t seed,
+                                  int scale = 10) {
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: size_t scale

File: tests/unittests/quantizationTest.cpp [hunk 3]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+    
+  }
+  // No-op TopK selection to test quantization
+  Node *downsample = F->createTopK("gru.downsample", hidden, EMBEDDING_SIZE/2);
+  
+  SaveNode *SN = F -> createSave("save", {downsample,0});
+  return {F, SN};
+}
+
+TEST_P(Quantization, end2endGRU) {
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: how long does it take to run the test?

File: tests/unittests/quantizationTest.cpp [hunk 4]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+					       {BATCH_SIZE,SEQUENCE_SIZE},
+					       "input", VisibilityKind::Public,
+					       Variable::TrainKind::None);
+  fillStableRandomIndex(input->getHandle<size_t>(), 7227, 10);
+
+  auto *hiddenInit = F->getParent()->createVariable(ElemKind::FloatTy,
+						    {BATCH_SIZE,EMBEDDING_SIZE},
+						    "hiddenInit", VisibilityKind::Public,
+						    Variable::TrainKind::None);
+  //fillStableRandomData(hidden->getHandle(), 3443, 1);
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: why is this commented out? If not needed, just remove.

File: tests/unittests/quantizationTest.cpp [hunk 5]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+
+  hiddenInit->getPayload().zero();
+
+  Node *hidden = hiddenInit;
+
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: can you clean those extra lines.

File: tests/unittests/quantizationTest.cpp [hunk 6]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+    H.raw(i) = int(i * 1921 + seed) % scale;
+  }
+}
+
+/// Builds a graph with two GRUs and saves output from last hidden node.
+static std::pair<Function *, SaveNode *>
+createGRUForQuantization(Module *M) {
+  Function *F = M->createFunction("main");
+
+  unsigned SEQUENCE_SIZE = 2;
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: can you make those constexpr, also for names, sequenceSize, embeddingSize etc.

File: tests/unittests/quantizationTest.cpp [hunk 7]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+
+  Node *hidden = hiddenInit;
+
+
+  for (unsigned step = 0; step < SEQUENCE_SIZE; step++){
+    // STEP2 - Gather a single set of embeddings for the GRU
+    Node *inputEmbedded = F->createGather("gru.embedding", emb, input);
+    Node *inputSlice = F->createSlice("gru.inputSlice",
+				      inputEmbedded,
+				      {0,step,0},
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: make sure to run clang-format on this file.

File: tests/unittests/quantizationTest.cpp [hunk 8]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+  EXPECT_EQ(result1Handle.size(), result2Handle.size());
+  
   for (int i = 0, e = result1Handle.size(); i < e; ++i) {
     float mx = result2Handle.raw(result2Handle.minMaxArg().second);
     double diff = std::fabs(result2Handle.raw(i) - result1Handle.raw(i)) / mx;
 
     // Allow 3% difference.
     EXPECT_NEAR(diff, 0, 0.03);
   }
+
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: kill this.

File: tests/unittests/quantizationTest.cpp (hunk 3 above)
Comment (rdzhabarov) [reply tests/unittests/quantizationTest.cpp]: when you run a coverage test, do you see quantization operator conversion code be fully covered now?

File: tests/unittests/quantizationTest.cpp [hunk 9]
```diff
@@ -216,14 +216,161 @@ TEST_P(Quantization, end2end) {
...
+				       F->createAdd("i_r_plus_h_r", i_r, h_r));
+    Node *inputgate = F->createSigmoid("gru.inputgate",
+				       F->createAdd("i_i_plus_h_i", i_i, h_i));
+    Node *newgate = F->createTanh("gru.newgate",
+				  F->createAdd("i_n_plus_rg_mult_h_n", i_n,
+					       F->createMul("rg_mult_h_n", resetgate, h_n)));
+    hidden = F->createAdd("gru.newhidden", newgate,
+			  F->createMul("ig_mult_hmng", inputgate,
+				       F->createSub("hidden_minus_newgate", hidden, newgate)));
+
```
Comment (rdzhabarov) [inline tests/unittests/quantizationTest.cpp]: kill these redundant lines

**Comment by rdzhabarov** [comment]
Looks great, please rebase based on the current master and address Nadav's comment.

File: tests/unittests/quantizationTest.cpp [hunk 10]
```diff
@@ -226,6 +225,141 @@ TEST_P(Quantization, end2end) {
...
+  SaveNode *SN = F->createSave("save", {downsample, 0});
+  return {F, SN};
+}
+
+TEST_P(Quantization, end2endGRU) {
+  auto res = createGRUForQuantization(&interpreterEE.getModule());
+  Function *F1 = res.first;
+  SaveNode *result1 = res.second;
+
+  glow::profileQuantization(F1);
```
[the remark this answers is not in the document]
Comment (rdzhabarov) [reply tests/unittests/quantizationTest.cpp]: Just follow the same route as it's in end2end quantization test (make sure to get the latest master changes).

**Comment by rdzhabarov** [comment]
LGTM, can you squash all your changes in one commit? and push it again?

**Comment by AnthonyOfSeattle** [comment]
Oh no, looks like some of my commits from a merge got included. Should I reopen a pull request to fix?

**Comment by AnthonyOfSeattle** [comment]
Squashed!

**Pull request merged**